### PR TITLE
TldParser update to v0.5.2, dotAnything for ANS, build fix

### LINKED ISSUE #2: [Bug] Error when trying to use the npm package in next js app
Hello when I try to add this package to my app i get the following error: 

error - .../app/node_modules/@portal-payments/solana-wallet-names/dist/index.js:1
import { PublicKey } from "@solana/web3.js";
^^^^^^

SyntaxError: Cannot use import statement outside a module
    at compileFunction (<anonymous>)
    at Object.compileFunction (node:vm:360:18)
    at wrapSafe (node:internal/modules/cjs/loader:1119:15)
    at Module._compile (node:internal/modules/cjs/loader:1155:27)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1245:10)
    at Module.load (node:internal/modules/cjs/loader:1069:32)
    at Function.Module._load (node:internal/modules/cjs/loader:904:12)
    at Module.require (node:internal/modules/cjs/loader:1093:19)
    at require (node:internal/modules/cjs/helpers:108:18)

### PR DESCRIPTION
**Pull request by tonivlad**
- Updated tldParser to v0.5.2 (the latest version works on react-native, and returns the real owner for Wrapped Domains)
- renamed the the ANS func from **dotAbcDotBonkOrDotPoorToWallet** -> **dotAnything**
- replaced esbuild + esbuild-plugin-d.t with tsup
  - it now generates cjs and mjs files
  - way faster
  - tsup is powered by esbuild (got the idea from https://www.npmjs.com/package/esbuild-plugin-d.ts)
- fixed some typescript
- bugfix: walletNameToAddressAndProfilePicture resolves .backpack owner even if there is no jwt

just saw that someone had the same error that I had soo 😂
this also fixes #2

### CODE REVIEW

**Comment by mikemaccana** [comment]
Thanks Vlad! The `tsup` stuff looks good + good catch on the .jwt.

We actually have some coding guidelines that I should have added to the README https://github.com/portalpayments/portalwallet/blob/main/CODING_GUIDELINES.md

Particularly:

1. https://github.com/portalpayments/portalwallet/blob/main/CODING_GUIDELINES.md#use-null-to-indicate-absense-of-a-value - get rid of all the explicit `undefined`s. It's an unrelated change (good PRs focus on  single thing), we don't want them, and it'll make the code review much easier.

2. Functions should do things. `dotAnythingWallet` doesn't make sense. It should `dotAnythingToWallet` like the other functions, or maybe dotAnythingToWalletAddress` (and update the other function names to match) to be more specific.

File: src/http-client.ts [hunk 1]
```diff
@@ -41,7 +41,7 @@ export const fetchUnfucked = async (
   }
   const response = await fetch(uri, options);
 
-  let contentType = CONTENT_TYPES.JSON;
+  let contentType: string | undefined = CONTENT_TYPES.JSON;
```
Comment (mikemaccana) [inline src/http-client.ts:44]: No `undefined`. It's always a string and it's default type is the content type for JSON.

File: src/index.test.ts [hunk 2]
```diff
@@ -12,9 +12,9 @@ import {
   dotSolToWallet,
   dotBackpackToWallet,
   dotGlowToWallet,
-  dotAbcDotBonkOrDotPoorToWallet,
+  dotAnythingToWallet,
   walletNameToAddressAndProfilePicture,
-  walletToDotAbcDotBonkOrDotPoor,
+  walletToDotAnything,
```
Comment (mikemaccana) [inline src/index.test.ts:17]: 👍🏻

File: src/index.test.ts [hunk 3]
```diff
@@ -287,7 +287,7 @@ describe(`wallet addresses to names`, () => {
     });
 
     test(`wallets with no names return null`, async () => {
-      const result = await walletToDotAbcDotBonkOrDotPoor(
+      const result = await walletToDotAnything(
         connection,
         WALLET_WITH_NO_NAMES
       );
```
Comment (mikemaccana) [inline src/index.test.ts:293]: Can you please add a new test for one of the wildcard domains? Like `a wallet address with an arbitrary domain registered at ANS returns the wallet name`?

File: src/index.ts [hunk 4]
```diff
@@ -19,13 +18,13 @@ const log = console.log;
 const stringify = (object: unknown) => JSON.stringify(object, null, 2);
 
 interface WalletNameAndProfilePicture {
-  walletName: string;
-  profilePicture: string;
+  walletName: string | null;
+  profilePicture: string | null;
```
Comment (mikemaccana) [inline src/index.ts:22]: Good catch 👍🏻

File: src/index.ts [hunk 5]
```diff
@@ -48,16 +47,16 @@ const removeExtension = (string: string, extension: string): string => {
...
+  ansDomainName: string
 ): Promise<WalletAddressAndProfilePicture> => {
   const parser = new TldParser(connection);
   const ownerPublicKey = await parser.getOwnerFromDomainTld(
-    dotAbcDotBonkOrDotPoorDomain
+    ansDomainName
   );
   return {
-    walletAddress: ownerPublicKey.toBase58(),
+    walletAddress: ownerPublicKey?.toBase58() || null,
```
Comment (mikemaccana) [inline src/index.ts:59]: 👍🏻

File: src/index.ts [hunk 6]
```diff
@@ -66,12 +65,12 @@ export const dotAbcDotBonkOrDotPoorToWallet = async (
...
 // https://github.com/onsol-labs/tld-parser/blob/main/tests/tld-parser.spec.ts#L97
 // getMainDomain() is what we want
-export const walletToDotAbcDotBonkOrDotPoor = async (
+export const walletToDotAnything = async (
   connection: Connection,
   wallet: PublicKey
 ): Promise<WalletNameAndProfilePicture> => {
   const parser = new TldParser(connection);
-  let mainDomain: MainDomain;
+  let mainDomain: any;
```
Comment (mikemaccana) [inline src/index.ts:73]: We generally try and avoid `any`. Should this be `string | null`?

File: src/index.ts [hunk 7]
```diff
@@ -237,7 +242,7 @@ export const dotBackpackToWallet = async (
   }
 
   const matchingUser = users.find(
-    (user) => user.username === dotBackpackUserName
+    (user: any) => user.username === dotBackpackUserName
```
Comment (mikemaccana) [inline src/index.ts:245]: Better to fix `users` (`null | Array<BackpackUser>`) - I can fix it if you want though.

File: src/index.ts [hunk 8]
```diff
@@ -258,7 +263,7 @@ export const dotBackpackToWallet = async (
     };
   }
 
-  const solanaPublicKeyDetails = publicKeysDetails.find((publicKeyDetails) => {
+  const solanaPublicKeyDetails = publicKeysDetails.find((publicKeyDetails: any) => {
```
Comment (mikemaccana) [inline src/index.ts:266]: Likewise, better to strictly type than to use any. But this is old code I can fix it if you like.

File: src/index.ts [hunk 9]
```diff
@@ -349,6 +354,7 @@ export const twitterHandleToWallet = async (
         profilePicture: null,
       };
     }
+    throw error;
```
Comment (mikemaccana) [inline src/index.ts:357]: Good catch!

File: src/index.ts [hunk 10]
```diff
@@ -418,6 +415,13 @@ export const walletNameToAddressAndProfilePicture = async (
       walletName
     );
   }
+  if (!walletAddressAndProfilePicture.walletAddress && walletName.split(".").length >= 2) {
```
Comment (mikemaccana) [inline src/index.ts:418]: Re `walletName.split(".").length >= 2)` - should this be a check earlier in the code, and we just return `null`? I don't know if any Solana wallet name type that is less than 2 parts.

File: tsup.config.ts [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+        "src/index.ts",
+        "src/types.ts",
+    ],
+    tsconfig: "tsconfig.json",
+    bundle: false,
+    splitting: false,
+    sourcemap: true,
+    clean: true,
+    dts: true
+})
```
Comment (mikemaccana) [inline tsup.config.ts:18]: 2 spaces (prettier should do this for you)

File: src/http-client.ts [hunk 12]
```diff
@@ -28,8 +28,8 @@ export const post = async (
 export const fetchUnfucked = async (
   uri: string,
   method = "GET",
-  headers: Record<string, string> = {},
-  body: Record<string, unknown> = null,
+  headers: Record<string, string> | null = {},
```
Comment (mikemaccana) [inline src/http-client.ts:31]: I don't think this needs to be `null`. Default headers are an empty object.

File: src/index.ts [hunk 13]
```diff
@@ -440,18 +444,18 @@ export const walletAddressToNameAndProfilePicture = async (
 ): Promise<WalletNameAndProfilePicture> => {
   const solanaPFPStandardImageURL =
     await getProfilePictureUsingSolanaPFPStandard(connection, wallet);
-  const dotAbcOrBonkOrPoor = await walletToDotAbcDotBonkOrDotPoor(
+  const dotAnything = await walletToDotAnything(
     connection,
     wallet
   );
   // .abc, .bonk and .poor service doesn't have a profile picture, so use Solana PFP Standard
```
Comment (mikemaccana) [inline src/index.ts:451]: Update the comment. It's no longer `.abc, .bonk and .poor service` just call it `ANS Domains`

**Review by mikemaccana — CHANGES_REQUESTED** [review]
Thanks for the update, it's now much easier to read! A few small fixes needed and this will be ready to merge!

File: src/index.ts (hunk 10 above)
Comment (tonivlad) [reply src/index.ts:418]: We have subdomains and so does Bonfida

though we don't expose creating a subdomain in the frontend, it is possible by calling the program directly

File: src/index.ts (hunk 13 above)
Comment (tonivlad) [reply src/index.ts]: updated

File: src/http-client.ts (hunk 1 above)
Comment (tonivlad) [reply src/http-client.ts:44]: it gets overwriten in line 50 and the .at(0) can return undefined

https://github.com/onsol-labs/solana-wallet-names/blob/7996db0a0b0e7b346a4e725402e926ac75c285a8/src/http-client.ts#L50

File: src/index.test.ts (hunk 3 above)
Comment (tonivlad) [reply src/index.test.ts:293]: can you give me more details on this, I don't quite understand what is needed

File: src/index.ts (hunk 6 above)
Comment (tonivlad) [reply src/index.ts:82]: we don't export the type and for the moment it could remain `any`

File: src/index.ts (hunk 7 above)
Comment (tonivlad) [reply src/index.ts:254]: I added an interface for BackpackUser but it's empty rn

 I don't know what it looks like, I don't know what to use for the jwt

File: src/index.ts (hunk 8 above)
Comment (tonivlad) [reply src/index.ts:275]: I added an interface for this but I don't know if it's right

I don't know what it looks like when the api url returns a full array (same as the previous jwt issue)

File: tsup.config.ts (hunk 11 above)
Comment (tonivlad) [reply tsup.config.ts:18]: fixed

File: src/http-client.ts (hunk 12 above)
Comment (tonivlad) [reply src/http-client.ts:31]: The `get` and `post` functions can send `null` as headers

File: src/http-client.ts (hunk 1 above)
Comment (tonivlad) [reply src/http-client.ts:44]: I added it so TS wouldn't not complain it being an error

File: src/index.ts (hunk 6 above)
Comment (tonivlad) [reply src/index.ts:82]: I'm going to update the https://www.npmjs.com/package/@onsol/tldparser lib today so we can go back to using MainDomain

Comment (tonivlad) [reply src/index.ts]: updated the tldParser lib it now exports the MainDomain type so this is fixed now

File: src/index.test.ts (hunk 3 above)
Comment (mikemaccana) [reply src/index.test.ts:293]: Handled in our call

File: src/index.ts (hunk 10 above)
Comment (mikemaccana) [reply src/index.ts]: Done in the call

**Review by mikemaccana — APPROVED** [review]
Thanks Vlad! Merged. 🚀

**Pull request merged**